### PR TITLE
Support define of variable-like preprocessor definitions by user

### PR DESCRIPTION
**Pull request by seri14**
for proposal

setifdef

### CODE REVIEW

**Comment by crosire** [comment]
I like this! Neat idea of using `#ifdef`/`#ifndef` to check for possible defines. My only concern is that yet another button above each effect is starting to make the UI a bit confusing. Maybe it makes sense to show them inline above the uniforms, not sure, will experiment a bit (my time is limited in the next few weeks though, so it might take me a while).

**Comment by CeeJayDK** [comment]
I'm Commander Shepard, and this is my favourite patch on Github.

**Comment by seri14** [comment]
smaa

File: source/runtime.cpp [hunk 1]
```diff
@@ -247,10 +247,16 @@ void reshade::runtime::load_effect(const std::filesystem::path &path, size_t &ou
...
+		if (pp.append_file(path))
+		{
+			effect.user_definitions = std::move(pp.user_definitions());
+		}
+		else
 		{
 			LOG(ERROR) << "Failed to load " << path << ":\n" << pp.errors();
 			effect.compile_sucess = false;
+
+			effect.user_definitions.clear();
```
Comment (crosire) [inline source/runtime.cpp]: This is not necessary. `effect` is a new object, so `user_definitions` will always be empty here anyway. There is nothing to clear. Either way, doesn't matter much, just do this for simplicity:
```c++
if (!pp.append_file(path))
{
...
}

effect.user_definitions = std::move(pp.user_definitions());
```

File: source/runtime.cpp [hunk 2]
```diff
@@ -273,12 +279,15 @@ void reshade::runtime::load_effect(const std::filesystem::path &path, size_t &ou
 			_renderer_id & 0x10000 ? reshadefx::codegen::backend::glsl : reshadefx::codegen::backend::hlsl;
 
 		// Compile the pre-processed source code (try the compile even if the preprocessor step failed to get additional error information)
-		if (!parser.parse(std::move(source_code), language, shader_model, true, _performance_mode, effect.module))
+		if (parser.parse(std::move(source_code), language, shader_model, true, _performance_mode, effect.module))
+		{
+			parser.errors();
```
Comment (crosire) [inline source/runtime.cpp]: This doesn't make sense? =)

File: source/gui.cpp [hunk 3]
```diff
@@ -1786,6 +1786,139 @@ void reshade::runtime::draw_overlay_variable_editor()
...
 			{
+				const float button_size = ImGui::GetFrameHeight();
+				const float button_spacing = _imgui_context->Style.ItemInnerSpacing.x;
+
+				ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, ImVec2(_imgui_context->Style.FramePadding.x, 0));
+				if (imgui_popup_button("Edit effect preprocessor definitions", -1, ImGuiWindowFlags_NoMove))
+				{
+					ImGui::PopStyleVar();
+
+					std::vector<std::string> user_definitions;
```
Comment (crosire) [inline source/gui.cpp]: This should be `std::set` to avoid duplicates (`set` instead of `unordered_set` to guarantee ordering in the UI).

**Comment by seri14** [comment]
This is draft. (I'm stopped work for proposal)
example, I guess it is required to support scrolling of definitions. (if use this)
and I do not know if it is good to filter definitions.
My idea is only exclude ReShade*.fxh

**Comment by seri14** [comment]
and behaviour when user mistaken set setting (filename is not have quote, etc)

**Comment by seri14** [comment]
back to self repository

**Pull request closed without merging**